File: src/main/kotlin/com/vlnabatov/alabaster/extensions/annotation/ScalaAnnotator.kt
package com.vlnabatov.alabaster.extensions.annotation

import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.lang.annotation.HighlightSeverity.INFORMATION
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.BRACES
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.NUMBER
import com.intellij.psi.PsiElement
import com.intellij.psi.util.elementType
import com.vlnabatov.alabaster.annotateSeparationMarks
import kotlin.collections.setOf
import org.jetbrains.plugins.scala.lang.lexer.ScalaTokenTypes.BOOLEAN_TOKEN_SET
import org.jetbrains.plugins.scala.lang.lexer.ScalaTokenTypes.kNULL
import org.jetbrains.plugins.scala.lang.lexer.ScalaTokenTypes.NUMBER_TOKEN_SET
import org.jetbrains.plugins.scala.lang.lexer.ScalaTokenTypes.STRING_LITERAL_TOKEN_SET


// Search in org.jetbrains.plugins.scala.lang.lexer.ScalaTokenTypes
class ScalaAnnotator : Annotator {
    override fun annotate(element: PsiElement, holder: AnnotationHolder) {
        try {
            when (element.elementType) {
                // constants
                in setOf(*BOOLEAN_TOKEN_SET.types, *NUMBER_TOKEN_SET.types, kNULL) ->
                    holder.newSilentAnnotation(INFORMATION).textAttributes(NUMBER).create()
                // strings
                in STRING_LITERAL_TOKEN_SET -> annotateSeparationMarks(element, holder, BRACES)
            }
        } catch (e: Exception) {
            /* Should not happen */
        }
    }
}

File: src/main/clojure/com/vlnabatov/alabaster/scalaannotator.clj
(ns com.vlnabatov.alabaster.scalaannotator
  (:require [com.vlnabatov.alabaster.util :refer [import-static]])
  (:import (com.intellij.lang.annotation AnnotationHolder)
           (com.intellij.lang.annotation HighlightSeverity)
           (com.intellij.openapi.editor DefaultLanguageHighlighterColors)
           (org.jetbrains.plugins.scala.lang.lexer ScalaTokenTypes)
           (com.intellij.psi PsiElement))
  (:gen-class :implements [com.intellij.lang.annotation.Annotator]
              :name com.vlnabatov.alabaster.extensions.annotation.CLJScalaAnnotator
              :prefix "-"))

;;(def val-tokens #{(.-types ScalaTokenTypes/BOOLEAN_TOKEN_SET)})

(defn -annotate
  [this ^PsiElement element ^AnnotationHolder holder]
  (let [elementType (.. element (getNode) (getElementType))
        is-string? (.contains ScalaTokenTypes/STRING_LITERAL_TOKEN_SET elementType)]
    (cond is-string? (-> holder
                         (.newSilentAnnotation HighlightSeverity/INFORMATION)
                         (.textAttributes DefaultLanguageHighlighterColors/MARKUP_ENTITY)
                         (.create))
          (.contains ScalaTokenTypes/STRING_LITERAL_TOKEN_SET elementType)
            (-> holder
                (.newSilentAnnotation HighlightSeverity/INFORMATION)
                (.textAttributes DefaultLanguageHighlighterColors/MARKUP_ENTITY)
                (.create)))))

